src/Elliptic/DiscontinuousGalerkin/DgElementArray.hpp

```diff
@@ -83,39 +87,67 @@ struct DefaultElementsAllocator
 
     const auto& blocks = domain.blocks();
 
-    const std::unordered_map<ElementId<Dim>, double> element_costs =
-        domain::get_element_costs(
-            blocks, initial_refinement_levels, initial_extents,
-            domain::ElementWeight::NumGridPoints, std::nullopt);
-    const domain::BlockZCurveProcDistribution<Dim> element_distribution{
-        element_costs,   num_of_procs_to_use, blocks, initial_refinement_levels,
-        initial_extents, procs_to_ignore};
+    const std::optional<domain::ElementWeight>& element_weight =
+        get<domain::Tags::ElementDistribution>(local_cache);
+
+    domain::BlockZCurveProcDistribution<Dim> element_distribution{};
+    if (element_weight.has_value()) {
+      const std::unordered_map<ElementId<Dim>, double> element_costs =
+          domain::get_element_costs(blocks, initial_refinement_levels,
+                                    initial_extents, element_weight.value(),
+                                    quadrature);
+      element_distribution = domain::BlockZCurveProcDistribution<Dim>{
+          element_costs,   num_of_procs_to_use,
+          blocks,          initial_refinement_levels,
+          initial_extents, procs_to_ignore};
+    }
 
     // Will be used to print domain diagnostic info
     std::vector<size_t> elements_per_core(number_of_procs, 0_st);
     std::vector<size_t> elements_per_node(number_of_nodes, 0_st);
     std::vector<size_t> grid_points_per_core(number_of_procs, 0_st);
     std::vector<size_t> grid_points_per_node(number_of_nodes, 0_st);
 
+    size_t which_proc = 0;
     for (const auto& block : blocks) {
       const size_t grid_points_per_element = alg::accumulate(
           initial_extents[block.id()], 1_st, std::multiplies<size_t>());
 
       const std::vector<ElementId<Dim>> element_ids = initial_element_ids(
           block.id(), initial_refinement_levels[block.id()]);
 
-      for (const auto& element_id : element_ids) {
-        const size_t target_proc =
-            element_distribution.get_proc_for_element(element_id);
-        element_array(element_id)
-            .insert(global_cache, initialization_items, target_proc);
-
-        const size_t target_node =
-            Parallel::node_of<size_t>(target_proc, local_cache);
-        ++elements_per_core[target_proc];
-        ++elements_per_node[target_node];
-        grid_points_per_core[target_proc] += grid_points_per_element;
-        grid_points_per_node[target_node] += grid_points_per_element;
+      if (element_weight.has_value()) {
+        for (const auto& element_id : element_ids) {
```

Comment:
Maybe add a comment here and in the else to the effect of `// Distribute by estimated cost of the element. E.g., a weighted space-filling curve.` and then `// Do a round-robin distribution of the elements`?